// cm6-graphql/dist/src/state.js
import { StateField, StateEffect } from '@codemirror/state';
const schemaEffect = StateEffect.define();
export const schemaStateField = StateField.define({
    create() { },
    update(schema, tr) {
        for (const e of tr.effects) {
            if (e.is(schemaEffect)) {
                return e.value;
            }
        }
        return schema;
    },
});
const optionsEffect = StateEffect.define();
export const optionsStateField = StateField.define({
    create() { },
    update(opts, tr) {
        for (const e of tr.effects) {
            if (e.is(optionsEffect)) {
                return e.value;
            }
        }
        return opts;
    },
});
export const updateSchema = (view, schema) => {
    view.dispatch({
        effects: schemaEffect.of(schema),
    });
};
export const updateOpts = (view, opts) => {
    view.dispatch({
        effects: optionsEffect.of(opts),
    });
};
export const getSchema = (state) => {
    return state.field(schemaStateField);
};
export const getOpts = (state) => {
    return state.field(optionsStateField);
};
const defaultOpts = {
    showErrorOnInvalidSchema: true,
};
export const stateExtensions = (schema, opts) => [
    schemaStateField.init(() => schema),
    optionsStateField.init(() => (Object.assign(Object.assign({}, defaultOpts), opts))),
];
